libmamba/include/mamba/util/iterator.hpp

```diff
@@ -0,0 +1,400 @@
+            if (n >= 0)
+            {
+                step_back(n);
+            }
+            else
+            {
+                advance(n);
+            }
+            return *this;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, difference_type>
+        operator-(const It& rhs) const
+        {
+            It tmp = rhs;
+            difference_type res = 0;
+            while (tmp++ != *this)
+            {
+                ++res;
+            }
+            return res;
+        }
+
+        reference operator*() const
+        {
+            return *m_iter;
+        }
+
+        pointer operator->() const
+        {
+            return m_iter.operator->();
+        }
+
+        bool equal(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter == rhs.m_iter;
+        }
+
+        template <class It = Iterator>
+        enable_random_access_iterator<It, bool>
+        less_than(const self_type& rhs) const
+        {
+            return m_pred == rhs.m_pred && m_iter < rhs.m_iter;
+        }
+
+    private:
+
+        void advance(difference_type n)
+        {
+            while (m_iter != m_end && n != 0)
+            {
+                ++(*this), --n;
+            }
+        }
+
+        void step_back(difference_type n)
+        {
+            while (m_iter != m_begin_limit && n != 0)
+            {
+                --(*this), --n;
+            }
+        }
+
+        void satisfy_predicate()
+        {
+            while (m_iter != m_end && !m_pred(*m_iter))
+            {
+                ++m_iter;
+            }
+        }
+
+        Predicate m_pred;
+        Iterator m_iter;
+        Iterator m_begin_limit;
+        Iterator m_end;
+    };
+
+    template <class Predicate, class Iterator>
+    bool operator==(const filter_iterator<Predicate, Iterator>& lhs,
+                    const filter_iterator<Predicate, Iterator>& rhs)
+    {
+        return lhs.equal(rhs);
+    }
+
+    template <class Predicate, class Iterator>
+    enable_random_access_iterator<Iterator, bool>
+    operator<(const filter_iterator<Predicate, Iterator>& lhs,
+              const filter_iterator<Predicate, Iterator>& rhs)
+    {
+        return lhs.less_than(rhs);
+    }
+
+    template <class Predicate, class Iterator>
+    filter_iterator<std::decay_t<Predicate>, Iterator>
+    make_filter_iterator(Predicate&& pred, Iterator iter, Iterator begin, Iterator end)
+    {
+        using return_type = filter_iterator<std::decay_t<Predicate>, Iterator>;
+        return return_type(
+            std::forward<Predicate>(pred),
+            iter,
+            begin,
+            end);
+    }
+
+    template <class Predicate, class C>
+    filter_iterator<std::decay_t<Predicate>, typename C::iterator>
+    filter_begin(Predicate&& pred, C& container)
+    {
+        using return_type = filter_iterator<std::decay_t<Predicate>, typename C::iterator>;
+        return return_type(
```

Comment:
You can actually remove the `return return_type(...);` expression and replace them by just
`return {...};`
This will also make sure that conversions are explicit, not implicit.